provider/kubernetes/kubernetes.go

```diff
@@ -116,45 +122,20 @@ func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *s
 	pool.Go(func(stop chan bool) {
 		operation := func() error {
 			for {
-				stopWatch := make(chan struct{}, 1)
-				defer close(stopWatch)
-				eventsChan, err := k8sClient.WatchAll(p.Namespaces, stopWatch)
+				err := p.watch(k8sClient, configurationChan)
 				if err != nil {
 					log.Errorf("Error watching kubernetes events: %v", err)
 					timer := time.NewTimer(1 * time.Second)
+					// Don't retry on a stop message, return nil.
 					select {
 					case <-timer.C:
 						return err
 					case <-stop:
```

Comment:
There is a problem with exiting the function if we want to exit by the stop signal, not an error

Reply:
I think @ZloyDyadka is right: The new method `watch()` isn't meant to return anymore, which means that we normally have no chance to process this line to terminate Ingress processing gracefully.